### PR TITLE
Fix `ChannelShuffle`

### PR DESCRIPTION
**Pull request by innat**
Close https://github.com/keras-team/keras-cv/issues/259

### CODE REVIEW

**Comment by bhack** [comment]
Was this done for `shufflenet`?

**Comment by innat** [comment]
I didn't test with the actual ShuffleNet model but I tested with this [model](https://github.com/keras-team/keras-cv/issues/259#issuecomment-1087768607), works fine so far. So, I think it should also work in general.

**Comment by bhack** [comment]
What I meant is that shufflenet  has not this randomizzation requirements so probably we need to do something different.

I don't think that we care about gradients in KLPs cause we have already many other ops  without gradient in KLPs so I don't think we are tracking this.

**Comment by innat** [comment]
> What I meant is that shufflenet has not this randomizzation requirements so probably we need to do something different.

Like this?

```python
tf.random.set_seed(self.seed)
tf.random.uniform(shape=[self.group], seed=self.seed)
```

**Comment by bhack** [comment]
Yes like:

https://github.com/tensorpack/tensorpack/blob/master/examples/ImageNetModels/shufflenet.py#L40-L49

**Comment by innat** [comment]
Yes. I found that earlier but couldn't adopt it because initially this layer was proposed as an input channel shuffle and randomness was a must.

However, now I think we can use `tf.random.set_seed`  to make it static or a one-time change for each channel layer's instance.

**Comment by bhack** [comment]
Let see if we could unify the two uses cases.

But It isn't clear to me if we want to reuse the KLP namespace in differentiable layers...

**Comment by bhack** [comment]
I think that we could separate Shuffle in Shufflenet and KLP layers as they have two very different behaviors.

Can we maintain the original impl for KLP?

File: keras_cv/layers/preprocessing/channel_shuffle.py [hunk 1]
```diff
@@ -65,10 +67,12 @@ def augment_image(self, image, transformation=None):
             )
 
         channels_per_group = num_channels // self.groups
-        image = tf.reshape(image, [height, width, self.groups, channels_per_group])
-        image = tf.transpose(image, perm=[2, 0, 1, 3])
-        image = tf.random.shuffle(image, seed=self.seed)
-        image = tf.transpose(image, perm=[1, 2, 3, 0])
+
+        rand_uniform = keras_cv.UniformFactorSampler(lower=0, upper=1, seed=self.seed)
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/channel_shuffle.py:71]: basically, you should do this in the constructor.

**Comment by LukeWood** [comment]
> Let's see if we could unify the two use cases.
> 
> But It isn't clear to me if we want to reuse the KLP namespace in differentiable layers...

I think it is fine.  Layers are imported from `keras_cv.layers.*`, where the code lives is less important.   ShuffleNet is a rare case.

File: keras_cv/layers/preprocessing/channel_shuffle.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/layers/preprocessing/channel_shuffle.py:71]: I understood your feedback earlier but recently I'm not in my workspace. I'll send update soon.

**Comment by LukeWood** [comment]
> I think that we could separate Shuffle in Shufflenet and KLP layers as they have two very different behaviors.
> 
> Can we maintain the original impl for KLP?

thanks for the note: can we clarify this point?  How does the behavior change?  This may be going over my head here

**Comment by bhack** [comment]
> > I think that we could separate Shuffle in Shufflenet and KLP layers as they have two very different behaviors.
> > 
> > Can we maintain the original impl for KLP?
> 
> thanks for the note: can we clarify this point?  How does the behavior change?  This may be going over my head here

What we are trying to fix with this PR? The missing gradient? avoid vmap fallback? XLA coverage? Something else?

**Comment by LukeWood** [comment]
> > > I think that we could separate Shuffle in Shufflenet and KLP layers as they have two very different behaviors.
> > > Can we maintain the original impl for KLP?
> > 
> > 
> > thanks for the note: can we clarify this point?  How does the behavior change?  This may be going over my head here
> 
> What we are trying to fix with this PR? The missing gradient? avoid vmap fallback? XLA coverage? Something else?

I imagine the missing gradient but that is a separate discussion imo.

What behavior mismatches here?  Functionally the layer does the same thing, right?

**Comment by innat** [comment]
@bhack 

> I think we could separate Shuffle in Shufflenet and KLP layers as they have two very different behaviors.

The operation will have a subtle difference if we want to maintain two separate implementations and the naming perhaps.

Also, currently, it uses two times transpose operation. But I think the new PR is more light. What do you think?

**Comment by innat** [comment]
> > What I meant is that shufflenet has not this randomizzation requirements so probably we need to do something different.
> 
> Like this?
> 
> ```python
> tf.random.set_seed(self.seed)
> tf.random.uniform(shape=[self.group], seed=self.seed)
> ```

@LukeWood 
Is it ok to use the above way?

**Comment by LukeWood** [comment]
> @bhack
> 
> > I think we could separate Shuffle in Shufflenet and KLP layers as they have two very different behaviors.
> 
> The operation will have a subtle difference if we want to maintain two separate implementations and the naming perhaps.
> 
> Also, currently, it uses two times transpose operation. But I think the new PR is more light. What do you think?

Can we clarify what the differences are?  I am  not convinced we need 2 implementations unless the differences are significant

**Comment by bhack** [comment]
> Can we clarify what the differences are?
Two points:
- I've [linked above](https://github.com/tensorpack/tensorpack/blob/master/examples/ImageNetModels/shufflenet.py#L40-L49) a quite popular TF implementation at that time and If I remember correctly from the paper (but I need to check it again) `channel_shuffle` has no random concept in ShuffleNet.
- Correct me if I am wrong. By an API point of view as our KLP layers are preprocessing layers they are not guaranteed do be differentiable. So it could be a little bit strange to use a layer from a KLP namespace in a "learnable" section of the network graph.

**Comment by innat** [comment]
@bhack 

> Correct me if I am wrong. By an API point of view as our KLP layers are preprocessing layers they are not guaranteed do be differentiable. So it could be a little bit strange to use a layer from a KLP namespace in a "learnable" section of the network graph.

I think it's fine to consider it a special case. The **static channel shuffle** operation is offered in pytorch officially but not as a **random layer** for image augmentation. But it does on the albumentaiton library. Either it treats as a differentiable layer or processing layer, in both case it has to subclass the `layer` module. [HERE](https://github.com/keras-team/keras-cv/pull/124#issuecomment-1042876194) was a relevant query. Also, I think the KPL layers may contain differentiable layers in the future.

**Comment by bhack** [comment]
> I think the KPL layers may contain differentiable layers in the future.

Isn't this point partially the root cause of this PR? If we are going in this direction we need to care every time about using or not differentiable ops when we contribute a new component in the `preprocessing` folder.
 
Your detection emerged with a sort of "e2e integration test" https://github.com/keras-team/keras-cv/pull/218#discussion_r840739049):
- we currently don't have this integration tests in the repo
- it emerged as we had a minor bug in TF for that specific op fixed by https://github.com/tensorflow/tensorflow/pull/55529.

**Comment by LukeWood** [comment]
> > I think the KPL layers may contain differentiable layers in the future.
> 
> Isn't this point partially the root cause of this PR? If we are going in this direction we need to care every time about using or not differentiable ops when we contribute a new component in the `preprocessing` folder.
> 
> Your detection emerged with a sort of "e2e integration test" [#218 (comment)](https://github.com/keras-team/keras-cv/pull/218#discussion_r840739049)):
> 
> * we currently don't have this integration tests in the repo
> * it emerged as we had a minor bug in TF for that specific op fixed by [Register Not differentiable RandomShuffle tensorflow/tensorflow#55529](https://github.com/tensorflow/tensorflow/pull/55529).

From what I can tell I think ShuffleNet is a super rare one off.

**Comment by bhack** [comment]
> From what I can tell I think ShuffleNet is a super rare one off.

Also If we consider this a quite rare case (I am not so future proof on this), for the API consistency we need to consider that this approach will require to remove the Random prefix from the name also if It has the (optional)  random logic. 
For all the other KLPs we used Random prefix  name when there was a random logic. 
Instead for this one, if we use the dual scope, we need to control its roandomicty by an extra param.
It think we are looping a Little bit of coherence.

> I think it's fine to consider it a special case. The static channel shuffle operation is offered in pytorch officially but not as a random layer for image augmentation

About pytorch there is a more general request of a random shuffle ops with axis param:

https://github.com/pytorch/pytorch/issues/71409

File: keras_cv/layers/preprocessing/channel_shuffle.py [hunk 2]
```diff
@@ -65,10 +72,10 @@ def augment_image(self, image, transformation=None):
             )
 
         channels_per_group = num_channels // self.groups
-        image = tf.reshape(image, [height, width, self.groups, channels_per_group])
-        image = tf.transpose(image, perm=[2, 0, 1, 3])
-        image = tf.random.shuffle(image, seed=self.seed)
-        image = tf.transpose(image, perm=[1, 2, 3, 0])
+
+        rand_indices = tf.argsort(self.rand_uniform(shape=[self.groups]))
```
Comment (innat) [inline keras_cv/layers/preprocessing/channel_shuffle.py:76]: @LukeWood cc. @bhack 
I've noticed that, the `tf.argsort` fallback to while loop here. 

```
Tensor("loop_body/argsort/TopKV2:1", shape=(3,), dtype=int32)
WARNING:tensorflow:Using a while_loop for converting TopKV2
```

File: keras_cv/layers/preprocessing/channel_shuffle.py [hunk 3]
```diff
@@ -52,6 +54,11 @@ def __init__(self, groups=3, seed=None, **kwargs):
         self.groups = groups
         self.seed = seed
 
+        tf.random.set_seed(self.seed)
```
[the remark this answers is not in the document]
Comment (bhack) [reply keras_cv/layers/preprocessing/channel_shuffle.py:57]: This is the point in my last comment

**Comment by innat** [comment]
@bhack cc. @LukeWood 

> Also If we consider this a quite rare case (I am not so future proof on this), for the API consistency we need to consider that this approach will require removing the Random prefix from the name also if It has the (optional) random logic.

IMO, I think it's not required to add a `random` prefix if we consider it as a special case. Curious, will `keras_cv` always provide only the random transformation layer? 

I'm ok to have two separate layers if it's accepted to `keras_cv`. But that will also feel a bit odd and naming might be confusing. 

I'm more interested to unify the two cases, i.e. for the random operation as KPL and for static operation as Model layers.

**Comment by bhack** [comment]
> Curious, will keras_cv always provide only the random transformation layer?

I think this part of the topic was early covered at https://github.com/keras-team/keras-cv/pull/122#discussion_r803214729. IMHO from that discussion we have not got any particular outcome.

File: keras_cv/layers/preprocessing/channel_shuffle.py (hunk 2 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/channel_shuffle.py:76]: > @LukeWood cc. @bhack I've noticed that, the `tf.argsort` fallback to while loop here.
> 
> ```
> Tensor("loop_body/argsort/TopKV2:1", shape=(3,), dtype=int32)
> WARNING:tensorflow:Using a while_loop for converting TopKV2
> ```

Just another one to add to the list. But as you can see with the master implementation we have also `RandomShuffle` in Keras-cv master (one in. one out):
https://github.com/keras-team/keras-cv/issues/291#issuecomment-1112210556

File: keras_cv/layers/preprocessing/channel_shuffle.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/channel_shuffle.py:57]: > @LukeWood cc @bhack
> 
> This line is only set to satisfy the shuffle-net's requirements. By this, it will produce a static shuffled operation.
> 
> ```python
> aug_fn = ChannelShuffle(groups=3, seed=101)
> aug_fn(a, training=True) 
> 
> aug_fn = ChannelShuffle(groups=3, seed=42)
> aug_fn(a, training=True) 
> ```
> 
> But for random operation, `seed=None`. But this may also be a bit in conflict with other KPLs from implementation perspective. Thoughts?

I don't think we should be modifying seed inside of any layer.

**Comment by LukeWood** [comment]
> > From what I can tell I think ShuffleNet is a super rare one off.
> 
> Also If we consider this a quite rare case (I am not so future proof on this), for the API consistency we need to consider that this approach will require to remove the Random prefix from the name also if It has the (optional) random logic. For all the other KLPs we used Random prefix name when there was a random logic. Instead for this one, if we use the dual scope, we need to control its roandomicty by an extra param. It think we are looping a Little bit of coherence.
> 
> > I think it's fine to consider it a special case. The static channel shuffle operation is offered in pytorch officially but not as a random layer for image augmentation
> 
> About pytorch there is a more general request of a random shuffle ops with axis param:
> 
> [pytorch/pytorch#71409](https://github.com/pytorch/pytorch/issues/71409)

Ok, yeah if there is no random behavior in one case and it is fully random in the other it should really be a different layer.  I misunderstood the original paper and meaning of ShuffleNet.

File: keras_cv/layers/preprocessing/channel_shuffle.py [hunk 4]
```diff
@@ -52,6 +54,11 @@ def __init__(self, groups=3, seed=None, **kwargs):
...
+        self.rand_uniform = keras_cv.UniformFactorSampler(
+            lower=0, upper=1, seed=self.seed
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/channel_shuffle.py:59]: can you add unit test for this gradient fix?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

**Comment by LukeWood** [comment]
Closing for now, as this does not seem to do the same thing as RandomChannelShuffle().  Instead; this appears to shuffle channels uniformly in order to support ShuffleNet.  If we ever support ShuffleNet, we can introduce a one-off layer to support that model architecture, in a gradient friendly approach.  

Thank you for your enthusiasm as always!
